vpr/src/base/partition.h

```diff
@@ -0,0 +1,52 @@
+#ifndef PARTITION_H
+#define PARTITION_H
+
+#include "vtr_strong_id.h"
+#include "region.h"
+#include "partition_regions.h"
+
+/**
+ * @file
+ * @brief This file defines the data for a partition: a grouping of atoms that are constrained to a portion of an FPGA. A partition defines
+ * the atoms that are assigned to it, and the locations in which they can be placed. The locations are a union of rectangles in x, y,
+ * sub_tile space. Common cases include a single rectangle or a single x, y, sub_tile (specific location). More complex partitions
+ * with L, T or other shapes can be created with a union of multiple rectangles.
+ */
+
+//Type tag for PartitionId
+struct partition_id_tag;
+
+//A unique identifier for a partition
+typedef vtr::StrongId<partition_id_tag> PartitionId;
+
+class Partition {
+  public:
+    PartitionId get_partition_id();
+    void set_partition_id(PartitionId _part_id);
+
+    std::string get_name();
+    void set_name(std::string _part_name);
+
+    //add an id to the atom_blocks vector
+    void add_to_atoms(AtomBlockId atom_id);
```

Comment:
This feels like it should be `add_atom` or `add_atom_to_partition`.

The description of the method describes the contents of the method, which is not helpful.  If someone wanted to understand the implementation, they can read the code.  The method comment should describe the why more than the how.

Couple example questions that are relevant to this method:
 - What happens if the same atom_id is added twice?  What happens then?
 - What is the implication of many partitions with few atoms versus fewer partitions with many atoms?
 - What is the implication of an atom being added to two Partition?

Reply:
Question: Is there a use case for providing a range of atoms, versus one at a time?

Reply:
Good points.  Adding an atom a second time doesn't really make sense; we could check and not add it, but that may get slow with a vector.  Could have a validate() method in vpr_constraints.h that checks for conditions like this efficiently (same atom in a partition twice, same atom in two different partitions) and prints a useful error message.

I don't think having an atom in two partitions makes sense.  It will complicate the data structures and the use case isn't very clear to me.  It makes it possible to have wildcards that overlap, e.g.:
*alu* --> partition 1
*alu|foo --> partition 2
That would only make sense if partition 2 is constrained to a region that is a subset of partition 1.  So you could make cases where this kind-of makes sense, but documenting it is going to be pain and it makes more complex code for not much gain.

Instead we could say:
A:  assigning the same atom to two partitions is an error (and then you have to check during parsing)
B:  The last assignment to an atom wins (removes any prior assignment).  That means if you code some relatively loose and some relatively specific constraints, you would order them from loose to specific to take advantage of this behaviour.

Either A or B seems like a reasonable choice.  Sarah, I suggest trying such a constraint in Quartus to see what it does; we don't have to do the same thing but it's useful to know what Quartus chose in such a case.

Reply:
I want to be clear, my examples where not about the implementation, but about the documentation and intention behind how to use the method.  If I find the "add_to_atoms" method, and I want to call it, what is valid, what isn't, etc.  Reading the implementation won't tell me answers to these questions, especially if the validation is done on the complete set of data.  It is totally fine in the comment to say, "see this other method that maintains the following invariant, etc".

Reply:
Sure, I'll check how Quartus handles this. In the case of choosing option B, would we then specify to users in documentation that their constraints should go from more general to more specific?

Reply:
Sarah: yes, we'd document the behavior and why (no matter what is chosen). 
Keith: agreed, this should be documented for the function in the .h file. But your comment also made me realize that the implementation also hasn't really made a choice on this; it's just letting the callers do whatever they want in terms of multiple additions of an atom. So we should both comment what invariants this function expects, and then enforce them either in the underlying add_atom function or in a validate function.